x-pack/plugins/task_manager/server/task_claimers/strategy_mget.ts

```diff
@@ -156,24 +161,61 @@ async function claimAvailableTasks(opts: TaskClaimerOpts): Promise<ClaimOwnershi
       continue;
     }
   }
+
   // apply limited concurrency limits (TODO: can currently starve other tasks)
   const candidateTasks = applyLimitedConcurrency(currentTasks, batches);
 
-  // build the updated task objects we'll claim
-  const taskUpdates: ConcreteTaskInstance[] = Array.from(candidateTasks)
-    .slice(0, initialCapacity)
-    .map((task) => {
-      if (task.retryAt != null && new Date(task.retryAt).getTime() < Date.now()) {
-        task.scheduledAt = task.retryAt;
+  // apply capacity constraint to candidate tasks
+  const tasksToRun: ConcreteTaskInstance[] = [];
+  const leftOverTasks: ConcreteTaskInstance[] = [];
+
+  let capacityAccumulator = 0;
+  for (const task of candidateTasks) {
+    const taskCost = definitions.get(task.taskType)?.cost ?? TaskCost.Normal;
+    if (capacityAccumulator + taskCost <= initialCapacity) {
+      tasksToRun.push(task);
+      capacityAccumulator += taskCost;
+    } else {
+      leftOverTasks.push(task);
+      capacityAccumulator = initialCapacity;
+    }
+  }
+
+  // perform an mget to get the full task instance for claiming
```

Comment:
Updated in https://github.com/elastic/kibana/pull/187999/commits/387cddfce46b0a5034834f086ece15b45cf2ba2b. In order to minimize type changes, I added a flag to the bulk update options to avoid overwriting the params and state fields with empty objects. I think this can be improved on followup as you suggested by switching to our own true partial update function.

Reply:
Makes sense to me, thanks!